ls1intum/Artemis
# `Exam mode`: Add initial dashboard for exam live statistics

### Checklist
#### General


#### Client

### Motivation and Context


As part of my thesis, we also want to introduce the client dashboard for the exam monitoring. This pull request presents the first initial version. This version will be updated and extended in several further pull requests.

### Description

Don't be afraid of the amount of new code added - there are also many parameterized tests :D 

- To view the dashboard, you need to enable exam monitoring and ...
  - ... go to the `Exam Detail` page and see the newly added section.
   ![image](https://user-images.githubusercontent.com/60005702/175230861-338706ed-bd85-4092-b451-ea0c11cca27b.png)
- The Dashboard contains multiple (sub)sections.
  - Overview: Shows the most important information of each subsection (clickable header, real-time charts, ...)
    - The exercise bar charts are updated every 5 seconds (to avoid too many re-renders)
    - The action line charts are updated every 10 seconds (to avoid too many re-renders)
  - Exercise/Activity-log: Will be implemented in a follow up (since this PR is already too big)
  - ...
- Each action performed by the student is sent to the server via websockets and then sent to the exam monitoring dashboard via websockets. As you can see, we have two functions which updates the data of each chart. At the moment, both functions behave very similarly. This will be changed in a follow-up to improve performance and to allow a reload/initial load.
  - `initData`: This functions is called on the initial load of the chart. We receive all performed actions and prepare them for the chart (implemented in a follow up).
  - `updateData`: This function is called on every update. It should only update the chart and not perform any data preparation (since we don't get any initial data at the moment, this function behaves like initData and will be adjusted in a follow up).

**Future PRs:**
In order to get this PR merged (and since this PR is already really large), we move some topics into additional PRs.
- Add initial loading
- Improve performance (if required)
- Add additional sections (exercises, activity log, ...)
- Improve tooltips
- Fine-tuning of actions
- ...

### Steps for Testing


Prerequisites:
- 1 Instructor
- Multiple Students
- 1 Exam with multiple exercises

1. Log in to Artemis
2. Navigate to Course Administration
3. Create an Exam with multiple Exercises and multiple registered Students
4. View the Exam Monitoring Dashboard as Instructor during the exam
5. Participate with multiple students during the exam
6. Verify that the collected actions are presented in the Dashboard

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| exam-monitoring-websocket.service.ts | 88.88% | 100% |
| exam-monitoring.component.ts | 87.5% | 100% |
| exam-monitoring.service.ts | 88.23% | 100% |
| chart.title.component.ts | 100% | 100% |
| monitoring-chart.ts | 83.87% | 97.5% |
| actions-chart.component.ts | 100% | 100% |
| average-actions-chart.component.ts | 75% | 94.73% |
| category-actions-chart.component.ts | 75% | 96.87% |
| total-actions-chart.component.ts | 75% | 95.23% |
| exercise-chart.component.ts | 100% | 93.75% |
| exercise-group-chart.component.ts | 87.5% | 95.23% |
| exercise-navigation-chart.component.ts | 100% | 95.45% |
| exercise-submission-chart.component.ts | 100% | 95.45% |
| exercise-template-chart.component.ts | 100% | 100% |
| monitoring-card.component.ts | 100% | 100% |
| monitoring-overview.component.ts | 75.67% | 100% |

### Screenshots



<details>
<summary>Exam Monitoring Dashboard (without data)</summary>

![image](https://user-images.githubusercontent.com/60005702/175283458-86fc6f16-69f7-4074-85be-7320e2f3def8.png)
![image](https://user-images.githubusercontent.com/60005702/175283504-61f47fd6-cc76-4f10-a787-ef8cf1259e35.png)
![image](https://user-images.githubusercontent.com/60005702/175283528-a25680b9-0396-4518-9587-66168d230054.png)
![image](https://user-images.githubusercontent.com/60005702/175283545-db4d7669-7fd8-4866-b5cc-b4e3e7640566.png)
![image](https://user-images.githubusercontent.com/60005702/175283581-c997caeb-69ee-4e1c-9e84-002287573d09.png)
![image](https://user-images.githubusercontent.com/60005702/175283618-e4d20764-61ae-43b8-bea8-d4fef526f176.png)
</details>

<details>
<summary>Exam Monitoring Dashboard (connected to server) - outdated</summary>

![Screenshot_20220610_132841](https://user-images.githubusercontent.com/60005702/173055644-118cbe34-f46d-42f8-9d02-0531ece6bfe4.png)
![image](https://user-images.githubusercontent.com/60005702/173055695-14fadc2e-a713-4c23-80e0-938d424c3014.png)

</details>
